File: S84.CTCode.Main.ctcode.cpp
#include "S84.CTCode.Main.ctcode.hpp"

namespace s84 {
namespace ctcode {
namespace main {
namespace ctcode {

Main::Main()
{
}

OmniPointer<s84::ctcode::transpiler::ctcode::Transpiler> Main::GetCPPTranspiler()
{
    OmniPointer<s84::ctcode::transpiler::cpptranspiler::ctcode::CPPTranspiler> specific = std::shared_ptr<s84::ctcode::transpiler::cpptranspiler::ctcode::CPPTranspiler>(new s84::ctcode::transpiler::cpptranspiler::ctcode::CPPTranspiler());
    specific->Initialize();
    OmniPointer<s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure> structue = std::shared_ptr<s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure>(new s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure());
    structue->SetTargetSpecificFunctions(specific);
    return structue;
}

OmniPointer<s84::ctcode::transpiler::ctcode::Transpiler> Main::GetPython3Transpiler()
{
    OmniPointer<s84::ctcode::transpiler::python3transpiler::ctcode::Python3Transpiler> specific = std::shared_ptr<s84::ctcode::transpiler::python3transpiler::ctcode::Python3Transpiler>(new s84::ctcode::transpiler::python3transpiler::ctcode::Python3Transpiler());
    specific->Initialize();
    OmniPointer<s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure> structue = std::shared_ptr<s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure>(new s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure());
    structue->SetTargetSpecificFunctions(specific);
    return structue;
}

OmniPointer<s84::ctcode::transpiler::ctcode::Transpiler> Main::GetNodeJSTranspiler()
{
    OmniPointer<s84::ctcode::transpiler::nodejstranspiler::ctcode::NodeJSTranspiler> specific = std::shared_ptr<s84::ctcode::transpiler::nodejstranspiler::ctcode::NodeJSTranspiler>(new s84::ctcode::transpiler::nodejstranspiler::ctcode::NodeJSTranspiler());
    specific->Initialize();
    OmniPointer<s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure> structue = std::shared_ptr<s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure>(new s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure());
    structue->SetTargetSpecificFunctions(specific);
    return structue;
}

OmniPointer<s84::ctcode::transpiler::ctcode::Transpiler> Main::GetJava11Transpiler()
{
    return std::shared_ptr<s84::ctcode::transpiler::java11transpiler::ctcode::Java11Transpiler>(new s84::ctcode::transpiler::java11transpiler::ctcode::Java11Transpiler());
}

OmniPointer<s84::ctcode::transpiler::ctcode::Transpiler> Main::GetLogToConsole()
{
    OmniPointer<s84::ctcode::transpiler::logtoconsole::ctcode::LogToConsole> specific = std::shared_ptr<s84::ctcode::transpiler::logtoconsole::ctcode::LogToConsole>(new s84::ctcode::transpiler::logtoconsole::ctcode::LogToConsole());
    specific->Initialize();
    OmniPointer<s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure> structue = std::shared_ptr<s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure>(new s84::ctcode::transpiler::standardstructure::ctcode::StandardStructure());
    structue->SetTargetSpecificFunctions(specific);
    return structue;
}

int Main::RunMain(OmniPointer<s84::ctcode::system::ctcode::System> system, std::string ctcode_file_name, std::string transpiler)
{
    OmniPointer<s84::ctcode::system::ctcode::OutputStream> logger = system->GetLoggerDestination();
    std::unordered_map<std::string, OmniPointer<s84::ctcode::transpiler::ctcode::Transpiler>> transpilers;
    SetKV(transpilers,std::string("CPPTranspiler"),this->GetCPPTranspiler());
    SetKV(transpilers,std::string("Python3Transpiler"),this->GetPython3Transpiler());
    SetKV(transpilers,std::string("NodeJSTranspiler"),this->GetNodeJSTranspiler());
    SetKV(transpilers,std::string("Java11Transpiler"),this->GetJava11Transpiler());
    SetKV(transpilers,std::string("LogToConsole"),this->GetLogToConsole());
    if (ctcode_file_name==std::string("")||!HasKV(transpilers,transpiler))
    {
        logger->WriteLine(std::string("ctcode <CTCodeFile> <Transpiler>"));
        logger->WriteLine(std::string("Known transpilers:"));
        std::vector<std::string> registered_transpilers = Keys(transpilers);
        int index = 0;
        while (index<Size(registered_transpilers))
        {
            logger->WriteLine(Concat(std::string("    "),Element(registered_transpilers,index)));
            index = index+1;
        }
        return 1;
    }
    std::string dbnf = system->ReadFileToString(ctcode_file_name);
    if (dbnf==std::string(""))
    {
        logger->WriteLine(Concat(std::string("The file "),Concat(ctcode_file_name,std::string(" is empty or does not exist."))));
        return 1;
    }
    OmniPointer<s84::ctcode::dbnf::ctcode::LargeString> dbnf_large_string = std::shared_ptr<s84::ctcode::dbnf::ctcode::LargeString>(new s84::ctcode::dbnf::ctcode::LargeString());
    dbnf_large_string->SetData(dbnf);
    OmniPointer<s84::ctcode::dbnf::ctcode::LengthString> index = std::shared_ptr<s84::ctcode::dbnf::ctcode::LengthString>(new s84::ctcode::dbnf::ctcode::LengthString());
    index->SetData(dbnf_large_string);
    index->SetStart(0);
    index->SetLength(Length(dbnf));
    OmniPointer<s84::ctcode::dbnf::ctcode::ParserNetwork> parser_network = std::shared_ptr<s84::ctcode::dbnf::ctcode::ParserNetwork>(new s84::ctcode::dbnf::ctcode::ParserNetwork());
    parser_network->Initialize();
    OmniPointer<s84::ctcode::dbnf::ctcode::CTCodeFileResult> ctcode_file_result = std::shared_ptr<s84::ctcode::dbnf::ctcode::CTCodeFileResult>(new s84::ctcode::dbnf::ctcode::CTCodeFileResult());
    OmniPointer<s84::ctcode::dbnf::ctcode::CTCodeFileParser> ctcode_file_parser = parser_network->GetCTCodeFileParser();
    logger->WriteLine(std::string("Parsing CTCodeFile..."));
    ctcode_file_parser->ParseSingleSave(index,ctcode_file_result);
    if (ctcode_file_result->GetResult()&&index->GetLength()==0)
    {
        logger->WriteLine(std::string("Done Parsing CTCodeFile!"));
        OmniPointer<s84::ctcode::transpiler::ctcode::Transpiler> target_transpiler = GetKV(transpilers,transpiler);
        return target_transpiler->Transpile(system,ctcode_file_result->GetValue(),ctcode_file_name);
    }
    else
    {
        logger->WriteLine(Concat(std::string("Failed to parse "),Concat(ctcode_file_name,std::string("."))));
        return 1;
    }
}


};
};
};
};

File: S84.CTCode.Main.ctcode.php
<?php
namespace S84\CTCode\Main\ctcode;

include_once "S84.CTCode.System.ctcode.php";
include_once "S84.CTCode.Transpiler.ctcode.php";
include_once "S84.CTCode.dbnf.ctcode.php";
include_once "S84.CTCode.Transpiler.StandardStructure.ctcode.php";
include_once "S84.CTCode.Transpiler.CPPTranspiler.ctcode.php";
include_once "S84.CTCode.Transpiler.CSharpTranspiler.ctcode.php";
include_once "S84.CTCode.Transpiler.Java11Transpiler.ctcode.php";
include_once "S84.CTCode.Transpiler.LogToConsole.ctcode.php";
include_once "S84.CTCode.Transpiler.NodeJSTranspiler.ctcode.php";
include_once "S84.CTCode.Transpiler.PHPTranspiler.ctcode.php";
include_once "S84.CTCode.Transpiler.Python3Transpiler.ctcode.php";
include_once "S84.CTCode.Transpiler.RubyTranspiler.ctcode.php";

function ClearList(array & $input): void { $input = array(); }
function Size(array $input): int { return count($input); }
function Element(array $input, int $element) { return $input[$element]; }
function Append(array & $input, mixed $element): void { $input[] = $element; }
function ClearMap(array & $input): void { reset($input); }
function SetKV(array & $input, string $key, mixed $element): void { $input[$key] = $element; }
function Keys(array $input): array { return array_keys($input); }
function HasKV(array $input, string $key): bool { return array_key_exists($key, $input); }
function GetKV(array $input, string $key) { return $input[$key]; }
function Length(string $input): int { return strlen($input); }
function At(string $input, int $index): string { return substr($input, $index, 1); }
function IntAt(string $input, int $index): int { return ord(substr($input, $index, 1)); }
function Concat(string $left, string $right): string { return $left . $right; }

class Main {
    public function __construct() {
    }

    public function GetCPPTranspiler(): ?\S84\CTCode\Transpiler\ctcode\Transpiler
    {
        $specific = new \S84\CTCode\Transpiler\CPPTranspiler\ctcode\CPPTranspiler();
        $specific->Initialize();
        $structue = new \S84\CTCode\Transpiler\StandardStructure\ctcode\StandardStructure();
        $structue->SetTargetSpecificFunctions($specific);
        return $structue;
    }

    public function GetCSharpTranspiler(): ?\S84\CTCode\Transpiler\ctcode\Transpiler
    {
        $specific = new \S84\CTCode\Transpiler\CSharpTranspiler\ctcode\CSharpTranspiler();
        $specific->Initialize();
        $structue = new \S84\CTCode\Transpiler\StandardStructure\ctcode\StandardStructure();
        $structue->SetTargetSpecificFunctions($specific);
        return $structue;
    }

    public function GetJava11Transpiler(): ?\S84\CTCode\Transpiler\ctcode\Transpiler
    {
        $specific = new \S84\CTCode\Transpiler\Java11Transpiler\ctcode\Java11Transpiler();
        $specific->Initialize();
        $structue = new \S84\CTCode\Transpiler\StandardStructure\ctcode\StandardStructure();
        $structue->SetTargetSpecificFunctions($specific);
        return $structue;
    }

    public function GetLogToConsole(): ?\S84\CTCode\Transpiler\ctcode\Transpiler
    {
        $specific = new \S84\CTCode\Transpiler\LogToConsole\ctcode\LogToConsole();
        $specific->Initialize();
        $structue = new \S84\CTCode\Transpiler\StandardStructure\ctcode\StandardStructure();
        $structue->SetTargetSpecificFunctions($specific);
        return $structue;
    }

    public function GetNodeJSTranspiler(): ?\S84\CTCode\Transpiler\ctcode\Transpiler
    {
        $specific = new \S84\CTCode\Transpiler\NodeJSTranspiler\ctcode\NodeJSTranspiler();
        $specific->Initialize();
        $structue = new \S84\CTCode\Transpiler\StandardStructure\ctcode\StandardStructure();
        $structue->SetTargetSpecificFunctions($specific);
        return $structue;
    }

    public function GetPHPTranspiler(): ?\S84\CTCode\Transpiler\ctcode\Transpiler
    {
        $specific = new \S84\CTCode\Transpiler\PHPTranspiler\ctcode\PHPTranspiler();
        $specific->Initialize();
        $structue = new \S84\CTCode\Transpiler\StandardStructure\ctcode\StandardStructure();
        $structue->SetTargetSpecificFunctions($specific);
        return $structue;
    }

    public function GetPython3Transpiler(): ?\S84\CTCode\Transpiler\ctcode\Transpiler
    {
        $specific = new \S84\CTCode\Transpiler\Python3Transpiler\ctcode\Python3Transpiler();
        $specific->Initialize();
        $structue = new \S84\CTCode\Transpiler\StandardStructure\ctcode\StandardStructure();
        $structue->SetTargetSpecificFunctions($specific);
        return $structue;
    }

    public function GetRubyTranspiler(): ?\S84\CTCode\Transpiler\ctcode\Transpiler
    {
        $specific = new \S84\CTCode\Transpiler\RubyTranspiler\ctcode\RubyTranspiler();
        $specific->Initialize();
        $structue = new \S84\CTCode\Transpiler\StandardStructure\ctcode\StandardStructure();
        $structue->SetTargetSpecificFunctions($specific);
        return $structue;
    }

    public function RunMain(?\S84\CTCode\System\ctcode\System $system, ?string $ctcode_file_name, ?string $transpiler): ?int
    {
        $logger = $system->GetLoggerDestination();
        $transpilers = array();
        SetKV($transpilers,'CPPTranspiler',$this->GetCPPTranspiler());
        SetKV($transpilers,'CSharpTranspiler',$this->GetCSharpTranspiler());
        SetKV($transpilers,'Java11Transpiler',$this->GetJava11Transpiler());
        SetKV($transpilers,'LogToConsole',$this->GetLogToConsole());
        SetKV($transpilers,'NodeJSTranspiler',$this->GetNodeJSTranspiler());
        SetKV($transpilers,'PHPTranspiler',$this->GetPHPTranspiler());
        SetKV($transpilers,'Python3Transpiler',$this->GetPython3Transpiler());
        SetKV($transpilers,'RubyTranspiler',$this->GetRubyTranspiler());
        if ($ctcode_file_name==''||!HasKV($transpilers,$transpiler))
        {
            $logger->WriteLine('ctcode <CTCodeFile> <Transpiler>');
            $logger->WriteLine('Known transpilers:');
            $registered_transpilers = Keys($transpilers);
            $registered_transpilers_index = 0;
            while ($registered_transpilers_index<Size($registered_transpilers))
            {
                $logger->WriteLine(Concat('    ',Element($registered_transpilers,$registered_transpilers_index)));
                $registered_transpilers_index = $registered_transpilers_index+1;
            }
            return 1;
        }
        $ctcode = $system->ReadFileToString($ctcode_file_name);
        if ($ctcode=='')
        {
            $logger->WriteLine(Concat('The file ',Concat($ctcode_file_name,' is empty or does not exist.')));
            return 1;
        }
        $ctcode_large_string = new \S84\CTCode\dbnf\ctcode\LargeString();
        $ctcode_large_string->SetData($ctcode);
        $index = new \S84\CTCode\dbnf\ctcode\LengthString();
        $index->SetData($ctcode_large_string);
        $index->SetStart(0);
        $index->SetLength(Length($ctcode));
        $parser_network = new \S84\CTCode\dbnf\ctcode\ParserNetwork();
        $parser_network->Initialize();
        $ctcode_file_result = new \S84\CTCode\dbnf\ctcode\CTCodeFileResult();
        $ctcode_file_parser = $parser_network->GetCTCodeFileParser();
        $logger->WriteLine('Parsing CTCodeFile...');
        $ctcode_file_parser->ParseSingleSave($index,$ctcode_file_result);
        if ($ctcode_file_result->GetResult()&&$index->GetLength()==0)
        {
            $logger->WriteLine('Done Parsing CTCodeFile!');
            $target_transpiler = GetKV($transpilers,$transpiler);
            return $target_transpiler->Transpile($system,$ctcode_file_result->GetValue(),$ctcode_file_name);
        }
        else
        {
            $logger->WriteLine(Concat('Failed to parse ',Concat($ctcode_file_name,'.')));
            return 1;
        }
    }

}

?>
